internal/bucket.go

```diff
@@ -0,0 +1,290 @@
+	// a parameter on the rule.
+	//
+	// This is for backwards-compatibility with older versions of the API,
+	// which didn't support setting shard-group durations and used an empty
+	// array of rules to represent infinite retention.
+	if rp > 0 || sgd > 0 {
+		rule := api.NewRetentionRuleWithDefaults()
+		rule.SetEverySeconds(int64(rp.Round(time.Second) / time.Second))
+		rule.SetShardGroupDurationSeconds(int64(sgd.Round(time.Second) / time.Second))
+		reqBody.RetentionRules = append(reqBody.RetentionRules, *rule)
+	}
+	if reqBody.OrgID == "" {
+		name := params.OrgName
+		if name == "" {
+			name = c.ActiveConfig.Org
+		}
+		lookupReq := clients.OrgApi.GetOrgs(ctx).Org(name)
+		resp, _, err := clients.OrgApi.GetOrgsExecute(lookupReq)
+		if err != nil {
+			return fmt.Errorf("failed to lookup ID of org %q: %w", name, err)
+		}
+		orgs := resp.GetOrgs()
+		if len(orgs) == 0 {
+			return fmt.Errorf("no organization found with name %q", name)
+		}
+		reqBody.OrgID = orgs[0].GetId()
+	}
+
+	req := clients.BucketApi.PostBuckets(ctx).PostBucketRequest(reqBody)
+	bucket, _, err := clients.BucketApi.PostBucketsExecute(req)
+	if err != nil {
+		return fmt.Errorf("failed to create bucket: %w", err)
+	}
+
+	return c.printBuckets(bucketPrintOptions{bucket: &bucket})
+}
+
+type BucketsListParams struct {
+	OrgID   string
+	OrgName string
+	Name    string
+	ID      string
+}
+
+func (c *CLI) BucketsList(ctx context.Context, client api.BucketsApi, params *BucketsListParams) error {
+	if params.OrgID == "" && params.OrgName == "" && c.ActiveConfig.Org == "" {
+		return errors.New("must specify org ID or org name")
+	}
+
+	req := client.GetBuckets(ctx)
+	if params.OrgID != "" {
+		req = req.OrgID(params.OrgID)
+	}
+	if params.OrgName != "" {
+		req = req.Org(params.OrgName)
+	}
+	if params.OrgID == "" && params.OrgName == "" {
+		req = req.Org(c.ActiveConfig.Org)
+	}
+	if params.Name != "" {
+		req = req.Name(params.Name)
+	}
+	if params.ID != "" {
+		req = req.Id(params.ID)
+	}
+
+	buckets, _, err := client.GetBucketsExecute(req)
+	if err != nil {
+		return fmt.Errorf("failed to list buckets: %w", err)
+	}
+
+	printOpts := bucketPrintOptions{}
+	if buckets.Buckets != nil {
+		printOpts.buckets = *buckets.Buckets
+	}
+	return c.printBuckets(printOpts)
+}
+
+type BucketsUpdateParams struct {
+	ID                 string
+	Name               string
+	Description        string
+	Retention          string
+	ShardGroupDuration string
+}
+
+func (c *CLI) BucketsUpdate(ctx context.Context, client api.BucketsApi, params *BucketsUpdateParams) error {
+	reqBody := api.PatchBucketRequest{}
+	if params.Name != "" {
+		reqBody.SetName(params.Name)
+	}
+	if params.Description != "" {
+		reqBody.SetDescription(params.Description)
+	}
+	if params.Retention != "" || params.ShardGroupDuration != "" {
+		patchRule := api.NewPatchRetentionRuleWithDefaults()
+		if params.Retention != "" {
+			rp, err := duration.RawDurationToTimeDuration(params.Retention)
```

Comment:
Is there any reason for parsing the string into a duration here rather than before the "top-level" `if` like on lines 34-41?

Reply:
Parsing collapses both `""` and `"0"` to the value `0`. We want to distinguish the two cases so users can explicitly update to a zero value, so the `if` checks `params.Retention` instead of the post-parsing value.

Reply:
Ah, nice!